### PR TITLE
Split output to stderr and stdout

### PR DESCRIPTION
**Pull request by SpacePossum**
For more correct testing I need https://github.com/symfony/symfony/pull/17255 to be resolved.

issue: https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1579 (and others)

### CODE REVIEW

**Comment by keradus** [comment]
What about switching to [SymfonyStyle](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/Style/SymfonyStyle.php) ?
`$io = new SymfonyStyle($input, $output);`

ref https://github.com/symfony/symfony/issues/15986

**Comment by GrahamCampbell** [comment]
We'd have to raise our min console vesion.

**Comment by keradus** [comment]
Raising dep version for just raising it is pointless. but rising it for functionality is OK ;)

**Comment by keradus** [comment]
BTW, then you won't need to test if console printer works on our end, just rely on vendor code (and it's tests)

**Comment by SpacePossum** [comment]
hmm even more work to be done in symfony/console... ;)

I don't like the current output construction, because if an output is given to a receiver/output writter it has to check if there is another output attached to it that is even more appropriate. This is causing code like this all over the place;
https://github.com/symfony/symfony/pull/15795/files#diff-425e56e3b866cb257f587805f7b2ee54R41
why isn't the error output passed to that method in the first place?
and why not set `hasErrOutput` on the `OutputInterface` that would at least remove the instance checks...?

anyway, `SymfonyStyle::error` still writes to stdout IIRC.. let me check the coming days :)

**Comment by keradus** [comment]
> why isn't the error output passed to that method in the first place?

For ensuring that if it is possible the stdErr is used.
If it would be added to method's prototype then it would break BC, and if not (last & nullable param) then nobody would pass it and stdErr would not be used, and it is not the goal. Bit ugliness for nice output out of the box without changing user code.

**Comment by keradus** [comment]
> anyway, `SymfonyStyle::error` still writes to stdout IIRC.. let me check the coming days :)

yes, and changing it will fix https://github.com/symfony/symfony/issues/15986 ;)

**Comment by keradus** [comment]
> We'd have to raise our min console version.

To use SymfonyStyle we will need to bump to v2.7, which has raised PHP requirement:
```diff
-        "php": ">=5.3.3"
+        "php": ">=5.3.9"
```

I'm ok with that, since it bump only PATCH version

**Comment by GrahamCampbell** [comment]
Maybe 2.x is a better place to make this change?

**Comment by keradus** [comment]
Could you justify that?

**Comment by SpacePossum** [comment]
>Could you justify that?

it would safe you the merge hell ;)

**Comment by keradus** [comment]
There should not be a lot of differences between 1.x and 2.x since the output of both should be the same, but it is fine for me.

**Comment by keradus** [comment]
What about creating our own Style on top of SymfonyStyle, so we could add more things there (like our progress) ?

**Comment by stof** [comment]
SymfonyStyle is about formatting the output. Telling that some styles must go to stdout and some other styles go to stderr is wrong. This is misunderstanding the goal of stderr and stdout (naming them stderr and stdout does not help understand their goal either btw).
any progress reporting output should also go to stderr, not to stdout (it is not part of the output of the program, which would be piped to another program)

**Comment by keradus** [comment]
Yep, we want to push the progress to stderr.
@stof, then what about https://github.com/symfony/symfony/issues/15986 ? Is it going to be the proposed way or the issue will be rejected ?

**Comment by keradus** [comment]
@SpacePossum I think we won't hear anything about this from Symfony soon.
Are we able to handle it on our side at low cost?

**Comment by SpacePossum** [comment]
>@SpacePossum I think we won't hear anything about this from Symfony soon.

Why do you think, all members are busy? I'll ping again, its been 10 days since my last proposal, hope I ping the right people ^-^
I could write a similar fix in the mean time for the fixer, but it would not be handy if my fix over at symfony gets turned down.

**Comment by keradus** [comment]
@SpacePossum , https://github.com/symfony/symfony/pull/15794 has been merged. Would it help us?

**Comment by SpacePossum** [comment]
@keradus that PR addresses some helpers we don't use ATM (and probably won't any time soon). 
I got an answer on 15986 from Fabien that contains enough pointer to finish it, I hope to find some time this weekend to do so.

**Comment by SpacePossum** [comment]
My PR to SF to get testing options for both stdout and stderr has been merged, however it has been done on 3.x line only. I would like to suggest to move on without it by adding a bit of a polyfill until the fixer requirements reach the 3.x line of packages of SF.
I would really like to wrap this one up to close the issues related to outputting to the wrong stream(s) making it harder than needed to integrate the fixer with other applications. I would like to suggest to rebase and rework this after #1824 has been done as the PRs will have some conflicts among each other. Ping @keradus for thoughts on this :)

**Comment by keradus** [comment]
fine for me

**Comment by keradus** [comment]
ping @SpacePossum

**Comment by SpacePossum** [comment]
I think a wait out #1824 since it will cover the output format testing. This means the tests here can be removed and this PR can focus on the used output streams only. I'll work on this soon to get the test logic in place we are going to need though :)

**Comment by keradus** [comment]
#1824 is merged. please update

File: Symfony/CS/Console/Command/FixCommand.php [hunk 1]
```diff
@@ -420,42 +420,42 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+                        $stdErr->writeln(sprintf('[%.3f s] %s', $event->getDuration() / 1000, $file));
                     }
 
-                    $output->writeln('');
+                    $stdErr->writeln('');
                 }
 
                 $fixEvent = $this->stopwatch->getEvent('fixFiles');
-                $output->writeln(sprintf('%s all files in %.3f seconds, %.3f MB memory used', $input->getOption('dry-run') ? 'Checked' : 'Fixed', $fixEvent->getDuration() / 1000, $fixEvent->getMemory() / 1024 / 1024));
+                $stdErr->writeln(sprintf('%s all files in %.3f seconds, %.3f MB memory used', $input->getOption('dry-run') ? 'Checked' : 'Fixed', $fixEvent->getDuration() / 1000, $fixEvent->getMemory() / 1024 / 1024));
```
Comment (GrahamCampbell) [inline Symfony/CS/Console/Command/FixCommand.php]: seems wrong to print information to stderr when it's not an error

Comment (SpacePossum) [inline Symfony/CS/Console/Command/FixCommand.php]: It is not, the naming of the streams is confusing. It is better, IMO, to send all kind of information to stdErr and only parsable formatted info to stdOut

**Comment by SpacePossum** [comment]
Reworked, please note this PR is now for ^1.12 only.

**Comment by keradus** [comment]
you mean that non of those changes should be merged into 2.x, right?

**Comment by SpacePossum** [comment]
>you mean that non of those changes should be merged into 2.x, right?

indeed, it is all good on 2.x :)

**Comment by keradus** [comment]
Cool ;)

**Comment by keradus** [comment]
Thank you @SpacePossum.

**Comment by TomasVotruba** [comment]
I'm writing post about CLI output styles PHP packages use and I Googled this is the 1 place in PHP CS Fixer that mentions `SymfonyStyle`. 

It was declided due to min-requiremeent bump to Symfony 2.8.

Now the [min. version is Symfony 3.2](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/1c10240da97479274fd40a136c3857ff94f7f93f/composer.json#L24), so it's ready :+1: 

Just FYI

**Pull request merged**